state/rethinkdb/rethinkdb.go

```diff
@@ -0,0 +1,364 @@
+	}
+
+	if c == nil {
+		return errors.Wrap(err, "invalid database response, cursor required")
+	}
+	defer c.Close()
+
+	var list []string
+	err = c.All(&list)
+	if err != nil {
+		return errors.Wrap(err, "invalid database responsewhile listing tables")
+	}
+
+	if !tableExists(list, stateTableName) {
+		_, err = r.DB(s.config.Database).TableCreate(stateTableName, r.TableCreateOpts{
+			PrimaryKey: stateTablePKName,
+		}).RunWrite(s.session)
+		if err != nil {
+			return errors.Wrap(err, "error creating state table in DB")
+		}
+	}
+
+	if s.config.Archive && !tableExists(list, stateArchiveTableName) {
+		// create archive table with autokey to preserve state id
+		_, err = r.DB(s.config.Database).TableCreate(stateArchiveTableName,
+			r.TableCreateOpts{PrimaryKey: stateArchiveTablePKName}).RunWrite(s.session)
+		if err != nil {
+			return errors.Wrap(err, "error creating state archive table in DB")
+		}
+		// index archive table for id and timestamp
+		_, err = r.DB(s.config.Database).Table(stateArchiveTableName).
+			IndexCreateFunc("state_index", func(row r.Term) interface{} {
+				return []interface{}{row.Field("id"), row.Field("timestamp")}
+			}).RunWrite(s.session)
+		if err != nil {
+			return errors.Wrap(err, "error creating state archive index in DB")
+		}
+	}
+
+	return nil
+}
+
+func tableExists(arr []string, table string) bool {
+	for _, a := range arr {
+		if a == table {
+			return true
+		}
+	}
+	return false
+}
+
+// Get retrieves a RethinkDB KV item
+func (s *RethinkDB) Get(req *state.GetRequest) (*state.GetResponse, error) {
+	if req == nil || req.Key == "" {
+		return nil, errors.New("invalid state request, missing key")
+	}
+
+	c, err := r.Table(stateTableName).Get(req.Key).Run(s.session)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting record from the database")
+	}
+
+	if c == nil || c.IsNil() {
+		return &state.GetResponse{}, nil
+	}
+
+	if c != nil {
+		defer c.Close()
+	}
+
+	var doc stateRecord
+	err = c.One(&doc)
+	if err != nil {
+		return nil, errors.Wrap(err, "error parsing database content")
+	}
+
+	resp := &state.GetResponse{ETag: doc.Hash}
+	b, ok := doc.Data.([]byte)
+	if ok {
+		resp.Data = b
+	} else {
+		data, err := json.Marshal(doc.Data)
+		if err != nil {
+			return nil, errors.New("error serializing data from database")
+		}
+		resp.Data = data
+	}
+	return resp, nil
+}
+
+// Set saves a state KV item
+func (s *RethinkDB) Set(req *state.SetRequest) error {
+	if req == nil || req.Key == "" || req.Value == nil {
+		return errors.New("invalid state request, key and value required")
+	}
+	return s.BulkSet([]state.SetRequest{*req})
+}
+
+// BulkSet performs a bulk save operation
+func (s *RethinkDB) BulkSet(req []state.SetRequest) error {
+	docs := make([]*stateRecord, len(req))
+	for i, v := range req {
+		docs[i] = &stateRecord{
+			ID:   v.Key,
+			TS:   time.Now().UTC().UnixNano(),
+			Hash: v.ETag,
+			Data: v.Value,
```

Comment:
how does rethinkdb client serialize value? Can you please double-check if this implementation has no same issue as [this](https://github.com/dapr/components-contrib/pull/460) ?

Reply:
As you know, that double serialization issue was global (as in Dapr-level) and not component specific. Also, the `With struct data` test covers that scenario.